### PR TITLE
PubChem importer

### LINKED ISSUE #2: 🧪 Create PubChem Database Importer Script
Please Read The WikiPages to complete script's requirements.

- [ ] [Overview](https://github.com/evrimagaci/periodum/wiki/Basic-Information,-Terms,-&-Goals)
- [ ] [Script's implementation](https://github.com/evrimagaci/periodum/wiki/Import-PubChem-Database)

### LINKED ISSUE #23: Switch from MySQL to PostgreSQL
Due to limited very types in MySQL (array type) we would like to switch to PostgreSQL.

### PR DESCRIPTION
**Pull request by T410**
This PR is to fetch Compounds from PubChem and strip them from unwanted data. 
See #2

### CODE REVIEW

**Review by T410 — COMMENTED** [review]
Please see the comment on `test/pubchem/import.spec.ts`

**Comment by T410** [comment]
I literally f*'d up by misclicking `review` 🤦

**Comment by T410** [comment]
Instead of creating a new PR, I hope people with write access can revert this back to a Draft PR 🙏

File: test/pubchem/import.spec.ts
Comment (T410) [inline test/pubchem/import.spec.ts]: Hey @nesimtunc I think I may need a little help here.

Line 3 and 9 are commented out since they break things. When importing `pubchem/index` and run `npm run test` I get this error `TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for`
and if I run `npm run test:watch`
I get this error `Error [ERR_REQUIRE_ESM]: Must use import to load ES Module:`

I think this is due to the `limiter` package not being CJS (not sure 🤔). Tried different things to resolve but no luck. I have thought about integrating the source code of `limiter` into the codebase (Their LICENSE is allowing it), modifying tsconfig.json (I don't want to break things for your end), tweaking test setup (not sure how to do that tbh)

Let me know how should I proceed
Thank you

**Comment by nesimtunc** [comment]
Hey @T410,

This PR looks interesting. I can see develop changes (you haven't merged our latest develop branch)? But also I see your commits. Is it formatting or what? This way it's difficult to see your changes.

Also, may I do the changes in your branch after you fix this?

**Comment by nesimtunc** [comment]
Hey @T410,

I have a  just fixed the conflicts on web.

File: test/pubchem/import.spec.ts
Comment (nesimtunc) [reply test/pubchem/import.spec.ts]: Please give me some time to look into this issue. I'll commit to this branch directly. @T410

**Comment by T410** [comment]
Hey @nesimtunc,

Sorry, I didn't have enough time to work on the issue. Thanks for the fix 🙌
I will try to find some time and work on writing tests and on prisma

**Comment by nesimtunc** [comment]
Hi @T410 

No worries, I have fixed only the conflicts, the test problem remains. I might have look into this weekend and probably will change to `jest`

Btw, no need to write tests for Prisma, just for PubChem script would be enough.

**Comment by nesimtunc** [comment]
Hi @T410 ,

Fixed the unit test problem, switched to `jest`, had to dealt with a lot of problems as you might guess, so, I couldn't write proper unit tests. 

Also, I have separated `api` (for data access) and service (for business logic) in pubchem, but haven't completed whole, rate limiter remains as a `// TODO` in `src/pubchem/service.ts`. Maybe you can refactor around, and write some tests.

That's all for now.

**Comment by nesimtunc** [comment]
Hi @T410,

We have decided to remove our compounds and use PubChem's compounds. I'm going to create a PR for that soon. Please merge it before importing the data. We'll need to use migration to update db schema.

**Comment by nesimtunc** [comment]
PR is ready now #22 to be review and merged

**Comment by T410** [comment]
Hey @nesimtunc,

I have written tests for service.test. But I am not sure I understood the reason for the import.test and specifically "should return related object when ToCHeading contains H1 key". 

I basically wrote the test for 'checking if the script is parsing the data correctly or not'. Not written tests for functions in parseData.ts

Let me know if that's correct or not.

Thanks

**Comment by nesimtunc** [comment]
Hi @T410 

Thank you very much for the tests, that's all nice. For the import part, we could check multiple import, or testing the throttle if possible, if not, you may delete it.

Besides that, inserting the data to the DB via Prisma, remains as the last task for this issue, right?

**Comment by T410** [comment]
Hey @nesimtunc,

I think throttle is testable. I will work on that. And yeah, I should also insert the data as well

**Comment by T410** [comment]
Hey @nesimtunc,

I know this issue was overdue and sorry about that. Hopefully, I have some good news. We are close to finishing this, I hope 🤞

### Updates:
No matter what I tried, I couldn't test the throttle logic 🙁

I have implemented the create logic but we should decide on either converting arrays into strings or creating new tables for them (I think that is called normalize?)

Also, after implementing everything I think @cagrimertbakirci should cross-check the data names and the spreadsheet that he created
https://docs.google.com/spreadsheets/d/1dH7d51UUrMOMmVlLZn6Mwk9zSIgJ3P5tp-V6edlqxQk/edit#gid=0

Let me know how should we approach the issue

Thanks a lot

Edit:
It's the first time I use Prisma and I may f'ed up the schema and or migrations. Please bare with me

**Comment by nesimtunc** [comment]
Hey @T410 

Thanks for the updates! We might switch to PostgreSQL to store `array` types which is also supported by Prisma. https://www.prisma.io/docs/concepts/database-connectors/postgresql#type-mapping-between-postgresql-to-prisma-schema

PostgreSQL can save even multi dimensional arrays. So, it might be better option.

**Comment by T410** [comment]
So, should I wait for the switch to continue working on this? Or ...?

Thanks

@nesimtunc

**Comment by nesimtunc** [comment]
Hi @T410,

I have just switched to PostgreSQL (it was easy with Prisma 🚀 ) and applied your migrations as well (array ones especially - whole schema) and it works fine. 

Rest is yours! 

Last but not least, I think now we need to talk about import process, we might change the strategy although we will still use your script. For sake of discussion it would be better to have a voice call, therefore I'd like to invite you to our Discord chat if that's okay with you?

Thanks again!

**Comment by T410** [comment]
Hey @nesimtunc,

Thanks for the response. Yeah, I'm ok with that. I think you can see my email in my profile for sending the invitation. Let's talk soon 🙌

**Comment by T410** [comment]
Hey @nesimtunc,

Just to let you know, as I try and go test other Compound ids more and more fields are turning into String Arrays. Almost every database field is string arrays now.

BTW, I have implemented compounds service and route. Will push a commit in a bit.
And fortunately either the limiter package that I'm using is not running correctly or PubChem API is not as reliable as I thought. I sometimes get 503 error from the API. I will try and tweak the limiter package and see if I still get as many errors as before.

Although I should implement a fallback logic for that too.

**Comment by nesimtunc** [comment]
Cool!

You may join the chat from this link, you can find me by my name :) 

https://discord.com/invite/evrimagaci

**Comment by T410** [comment]
I sent you a message on Discord

**Comment by T410** [comment]
Per our conversation with @nesimtunc, here are the tasks that need to be done

- [ ] Create a table in the DB for keeping track of the imported/failed compounds.
- [ ] Use that table to rerun the script if the script crashes unexpectedly.
- [ ] Limit retry amounts for failed compounds.
- [ ] Create an endpoint to show the statistics, ie: the time to import a number of compounds, success/fail ratio, how many compounds are left to import
- [ ] Create a POST endpoint to trigger the import script with an authentication(token may be stored in .env) header with a JSON body that includes the startID and endID and an option to continue from the last compound
- [ ] Create a POST endpoint to trigger pause and stop the import script

**Review by nesimtunc — APPROVED** [review]

**Pull request merged**
